src/modules/sdbus/message.c

```diff
@@ -0,0 +1,486 @@
+        case 'q':
+            n.u16 = json_integer_value (o);
+            break;
+        case 'i':
+            n.i32 = json_integer_value (o);
+            break;
+        case 'u':
+            n.u32 = json_integer_value (o);
+            break;
+        case 'x':
+            n.i64 = json_integer_value (o);
+            break;
+        case 't':
+            n.u64 = json_integer_value (o);
+            break;
+        case 'h':
+            n.i = json_integer_value (o);
+            break;
+        case 'd':
+            n.f = json_real_value (o);
+            break;
+        default:
+            return -EPROTO;
+    }
+    return sd_bus_message_append_basic (m, type, &n);
+}
+
+static int sdmsg_put_variant (sd_bus_message *m, json_t *o)
+{
+    const char *type;
+    json_t *val;
+    int e;
+
+    if (json_unpack (o, "[so]", &type, &val) < 0)
+        return -EPROTO;
+    if ((e = sd_bus_message_open_container (m, 'v', type)) < 0)
+        return e;
+    if ((e = sdmsg_put (m, type, val)) < 0)
+        return e;
+    if ((e = sd_bus_message_close_container (m)) < 0)
+        return e;
+    return 0;
+}
+
+static int sdmsg_put_struct (sd_bus_message *m, const char *fmt, json_t *o)
+{
+    int e;
+
+    if ((e = sd_bus_message_open_container (m, 'r', fmt)) < 0)
+        return e;
+    if ((e = sdmsg_write (m, fmt, o)) < 0)
+        return e;
+    if ((e = sd_bus_message_close_container (m)) < 0)
+        return e;
+    return 0;
+}
+
+int sdmsg_put (sd_bus_message *m, const char *fmt, json_t *o)
+{
+    int e;
+
+    if (streq (fmt, "a(sv)"))
+        e = sdmsg_put_array (m, "(sv)", o);
+    else if (streq (fmt, "a(sasb)"))
+        e = sdmsg_put_array (m, "(sasb)", o);
+    else if (fmt[0] == 'a' && strlen (fmt) == 2)
+        e = sdmsg_put_array (m, &fmt[1], o);
+    else if (streq (fmt, "(sv)"))
+        e = sdmsg_put_struct (m, "sv", o);
+    else if (streq (fmt, "(sasb)"))
+        e = sdmsg_put_struct (m, "sasb", o);
+    else if (streq (fmt, "v"))
+        e = sdmsg_put_variant (m, o);
+    else if (strlen (fmt) == 1)
+        e = sdmsg_put_basic (m, fmt[0], o);
+    else
+        e = -EPROTO;
+    return e;
+}
+
+int sdmsg_write (sd_bus_message *m, const char *fmt, json_t *o)
+{
+    int e;
+
+    if (!json_is_array (o))
+        return -EPROTO;
+
+    int cursor = 0;
+    for (int i = 0; fmt[i] != '\0';) {
+        json_t *entry;
+        char *efmt = NULL;
+
+        if (!(entry = json_array_get (o, cursor++)))
+            return -EPROTO;
+        if (!strncmp (&fmt[i], "a(sasb)", 7))
+            efmt = strndup (&fmt[i], 7);
+        else if (!strncmp (&fmt[i], "a(sv)", 5))
+            efmt = strndup (&fmt[i], 5);
+        else if (fmt[i] == 'a' && strlen (&fmt[i]) > 1)
+            efmt = strndup (&fmt[i], 2);
+        else
+            efmt = strndup (&fmt[i], 1);
+        if (!efmt)
+            return -ENOMEM;
+        if ((e = sdmsg_put (m, efmt, entry)) < 0) {
+            free (efmt);
+            return e;
+        }
```

Comment:
i think json_decref(entry) on error path?

Reply:
No, `entry` is an element of the function parameter `o` which belongs to the caller.  We use `json_array_get()` to access it which does not take a reference.